File: SWEA/D5/1248. ［S／W 문제해결 응용］ 3일차 － 공통조상/［S／W 문제해결 응용］ 3일차 － 공통조상.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Solution {

	static int V, E;
	static int v1, v2;
	static Node[] nodes;
	static boolean[] visited;
	static int size;

	public static void main(String[] args) throws Exception {
//		System.setIn(new FileInputStream("input.txt"));
		final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		final StringBuilder sb = new StringBuilder();
		StringTokenizer st;
		final int T = Integer.parseInt(br.readLine());
		for (int testCase = 1; testCase <= T; testCase++) {
			st = new StringTokenizer(br.readLine());
			V = Integer.parseInt(st.nextToken());
			E = Integer.parseInt(st.nextToken());
			v1 = Integer.parseInt(st.nextToken());
			v2 = Integer.parseInt(st.nextToken());
			nodes = new Node[V + 1];
			for (int i = 1; i <= V; i++) {
				nodes[i] = new Node(i);
			}
			st = new StringTokenizer(br.readLine());
			for (int i = 0; i < E; i++) {
				final int parent = Integer.parseInt(st.nextToken());
				final int child = Integer.parseInt(st.nextToken());
				if (nodes[parent].leftChild == 0) {
					nodes[parent].leftChild = child;
				} else {
					nodes[parent].rightChild = child;
				}
				nodes[child].parent = parent;
			}
			visited = new boolean[V + 1];
			Node v1Node = nodes[v1];
			Node v2Node = nodes[v2];
			Node commonParent = nodes[1];
			while (v1Node.data != 1) {
				visited[v1Node.data] = true;
				v1Node = nodes[v1Node.parent];
			}
			while (v2Node.data != 1) {
				if (visited[v2Node.data]) {
					commonParent = v2Node;
					break;
				}
				visited[v2Node.data] = true;
				v2Node = nodes[v2Node.parent];
			}
			size = 0;
			traversal(commonParent);
			sb.append("#")
					.append(testCase)
					.append(" ")
					.append(commonParent.data)
					.append(" ")
					.append(size)
					.append("\n");
		}
		System.out.println(sb);
	}

	private static void traversal(final Node node) {
		size++;
		if (node.leftChild != 0) {
			traversal(nodes[node.leftChild]);
		}
		if (node.rightChild != 0) {
			traversal(nodes[node.rightChild]);
		}
	}
}

class Node {

	final int data;
	int parent, leftChild, rightChild;

	public Node(final int data) {
		this.data = data;
	}
}

File: SWEA/D5/1248. ［S／W 문제해결 응용］ 3일차 － 공통조상/［S／W 문제해결 응용］ 3일차 － 공통조상.py
class Node:
    data = 0
    parent = 0
    left_child = 0
    right_child = 0

    def __init__(self, data):
        self.data = data


def traversal(node):
    global size
    size += 1
    if node.left_child != 0:
        traversal(nodes[node.left_child])
    if node.right_child != 0:
        traversal(nodes[node.right_child])


T = int(input())
for test_case in range(1, T + 1):
    V, E, v1, v2 = map(int, input().split())
    nodes = [Node(i) for i in range(V + 1)]
    links = list(map(int, input().split()))
    for i in range(E):
        p = links[2 * i]
        c = links[2 * i + 1]
        if nodes[p].left_child == 0:
            nodes[p].left_child = c
        else:
            nodes[p].right_child = c
        nodes[c].parent = p
    v1_node = nodes[v1]
    v2_node = nodes[v2]
    visited = [0] * (V + 1)
    common_parent = nodes[1]
    while v1_node.data != 1:
        visited[v1_node.data] = 1
        v1_node = nodes[v1_node.parent]
    while v2_node.data != 1:
        if visited[v2_node.data]:
            common_parent = v2_node
            break
        visited[v2_node.data] = 1
        v2_node = nodes[v2_node.parent]
    size = 0
    traversal(common_parent)
    print(f'#{test_case} {common_parent.data} {size}')
